app/gradle/journeyTest.gradle.kts

```diff
@@ -0,0 +1,100 @@
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+import batect.buildtools.JourneyTestNamingCheckTask
+
+val sourceSets = the<SourceSetContainer>()
+
+val journeyTest by sourceSets.creating {
+    compileClasspath += sourceSets["testCommon"].output
+    runtimeClasspath += sourceSets["testCommon"].output
+}
+
+configurations["journeyTestImplementation"].extendsFrom(configurations["testImplementation"])
+configurations["journeyTestRuntimeOnly"].extendsFrom(configurations["testRuntimeOnly"])
+
+dependencies {
+    "journeyTestImplementation"(libs.kotest.assertions.core)
+}
+
+abstract class JourneyTest : Test()
+
+tasks.withType<JourneyTest>().configureEach {
+    group = "Verification"
+    testClassesDirs = journeyTest.output.classesDirs
+    classpath = journeyTest.runtimeClasspath
+
+    dependsOn(tasks["journeyTestClasses"])
+    dependsOn(tasks["installShadowDist"])
+
+    inputs.files(tasks["installShadowDist"])
+
+    environment("BATECT_ENABLE_TELEMETRY", "false")
+}
+
+val windowsTests = "batect/journeytests/windowscontainers/**"
+val cleanupTests = "batect/journeytests/cleanup/**"
+
+val journeyTestWindowsContainers by tasks.registering(JourneyTest::class) {
+    description = "Runs the journey tests for Windows containers."
+
+    include(windowsTests)
+
+    onlyIf { "windowsContainers" in project.properties }
+}
+
+val journeyTestLinuxContainersParallelisable by tasks.registering(JourneyTest::class) {
+    description = "Runs the journey tests for Linux containers that can run in parallel."
+
+    exclude(windowsTests)
+    exclude(cleanupTests)
+
+    onlyIf { "windowsContainers" !in project.properties }
+
+    maxParallelForks = 4
+}
+
+val journeyTestLinuxContainersNonParallelisable by tasks.registering(JourneyTest::class) {
+    description = "Runs the journey tests for Linux containers that cannot run in parallel."
+
+    include(cleanupTests)
+
+    onlyIf { "windowsContainers" !in project.properties }
+}
+
+val journeyTestLinuxContainers by tasks.registering(JourneyTest::class) {
+    description = "Runs the journey tests for Linux containers."
+    group = "Verification"
+
+    dependsOn(journeyTestLinuxContainersParallelisable)
+    dependsOn(journeyTestLinuxContainersNonParallelisable)
+}
+
+tasks.register("journeyTest") {
+    description = "Runs the journey tests."
+    group = "Verification"
+
+    dependsOn(journeyTestWindowsContainers)
+    dependsOn(journeyTestLinuxContainers)
+}
+
+tasks.named<JourneyTestNamingCheckTask>("checkJourneyTestNaming").configure {
```

Comment:
@charleskorn I'm a bit clueless here as this line triggers

```
Task with name 'checkJourneyTestNaming' not found in project ':app'.
```

However, if I change the line to `tasks.regsiter<JourneyTestNamingCheckTask>("checkJourneyTestNaming") {` I get

```
A problem occurred configuring project ':app'.
> Cannot add task 'checkJourneyTestNaming' as a task with that name already exists.
```

So Gradle seems to be confused about whether the tasks already exists or not... 😄

Any idea how to solve this?

Reply:
Hmm, looks like wrapping this in `afterEvaluate` does the trick...

Reply:
As @vampire kindly pointed out, the reason for the need to use `afterEvaluate` here is tat the plugin itself gets only registered during `afterEvaluate`:

https://github.com/batect/batect/blob/08b33266bf9e05b9efbcda6b636a015bc6c6cd2f/buildSrc/src/main/kotlin/batect/buildtools/KotlinPlugin.kt#L183-L190

@charleskorn do you see a chance for fixing that first in the plugin? What's the reason for only registering it at `afterEvaluate`?

Reply:
I'll leave it to using `afterEvaluate` for now.

Reply:
I think the use of `afterEvaluate` was to work around a timing problem (the plugin is applied first, then this script, so the source set would never exist when the plugin is applied). I'll merge this as-is and see if I can re-jig the plugin to just react to the registration of the `journeyTest` source set.

Reply:
https://github.com/batect/batect/commit/1e3e823d381e7e7a436f3a8f4aea41a7040465d6 seems to work and removes the need for `afterEvaluate`.